ctera/src/main/java/ch/cyberduck/core/ctera/CteraCustomActionVersioning.java

```diff
@@ -0,0 +1,163 @@
+                                    "<input type=\"hidden\" name=\"ctera_ticket\" value=\"%s\"/>\n" +
+                                    "</form> </body> </html>",
+                            new HostUrlProvider().withUsername(false).get(session.getHost()),
+                            URLEncoder.encode(file.getAbsolute(), StandardCharsets.UTF_8.name()),
+                            token);
+            final Local file = TemporaryFileServiceFactory.get().create(String.format("%s.html", new AlphanumericRandomStringService().random()));
+            new DefaultLocalTouchFeature().touch(file);
+            try (final OutputStream out = file.getOutputStream(false)) {
+                IOUtils.write(content, out, StandardCharsets.UTF_8);
+            }
+            return file;
+        }
+        catch(IOException e) {
+            throw new DefaultIOExceptionMappingService().map("Error writing temporary file", e);
+        }
+    }
+
+    protected String getSessionToken() throws BackgroundException {
+        try {
+            final String device = this.getDeviceName();
+            if(log.isDebugEnabled()) {
+                log.debug(String.format("Using device %s to request a session token", device));
+            }
+            final HttpPost post = new HttpPost(String.format("/ServicesPortal/api/devices/%s?format=jsonext", device));
+            post.setEntity(new StringEntity(getSessionTokenPayloadAsString(), ContentType.APPLICATION_JSON));
+            return session.getClient().execute(post, response -> StringUtils.remove(EntityUtils.toString(response.getEntity()), "\""));
+        }
+        catch(IOException e) {
+            throw new DefaultIOExceptionMappingService().map(String.format("Unable to redirect to the portal for %s", file), e);
+        }
+    }
+
+    private String getDeviceName() throws BackgroundException {
+        final String token = this.getCteraTokens();
+        try {
+            if(StringUtils.isNotBlank(token)) {
+                final CteraTokens tokens = CteraTokens.parse(token);
+                final String id = tokens.getDeviceId();
+                final HttpGet request = new HttpGet("/ServicesPortal/api/devices?format=jsonext");
+                final Device[] devices = session.getClient().execute(request, new AbstractResponseHandler<Device[]>() {
+                    @Override
+                    public Device[] handleEntity(final HttpEntity entity) throws IOException {
+                        ObjectMapper mapper = new ObjectMapper();
+                        return mapper.readValue(entity.getContent(), Device[].class);
+                    }
+                });
+                for(Device device : devices) {
+                    if(device.uid.equals(id)) {
+                        return device.name;
+                    }
+                }
+            }
+            throw new InteroperabilityException("No token or device found");
```

Comment:
Throw _No token found in keychain_ when keychain returns null and _No device found_ when device array is empty.